scripts/cybershake/cybershake_progress.py

```diff
@@ -0,0 +1,364 @@
+        cur_fault_chours = np.asarray(
+            [
+                get_chours(json_file, PROCESS_TYPES)[1]
+                for json_file in fault.glob(f"**/{const.METADATA_LOG_FILENAME}")
+            ]
+        ).reshape(-1, len(PROCESS_TYPES))
+
+        faults.append(fault_name)
+        if cur_fault_chours.shape[0] > 0:
+            core_hours.append(np.nansum(cur_fault_chours, axis=0))
+        else:
+            core_hours.append([np.nan, np.nan, np.nan])
+
+    df = pd.DataFrame(index=faults, data=core_hours, columns=PROCESS_TYPES)
+
+    return df
+
+
+def get_completed_r_count(db_file, proc_type: str, fault_name: str = None):
+    """Returns the number of realisations of a fault "fault_name" that completed the proc_type. If fault_name is None,
+    consider all faults """
+    proc_type_int = const.ProcessType[proc_type].value
+    conn = sqlite3.connect(str(db_file))
+    c = conn.cursor()
+    if fault_name is None:
+        c.execute(
+            f"SELECT COUNT(*) FROM state WHERE proc_type=? and status={const.Status['completed'].value}",
+            (proc_type_int,),
+        )
+    else:
+        c.execute(
+            f"SELECT COUNT(*) FROM state WHERE proc_type=? and status={const.Status['completed'].value} and run_name LIKE ?",
+            (proc_type_int, fault_name + "_REL%"),
+        )
+
+    completed_r_count = c.fetchone()[0]
+    conn.close()
+    return completed_r_count
+
+
+def get_faults_and_r_count(cybershake_list: str):
+    """Gets the fault names and number of realisations from a cybershake fault list."""
+    with open(cybershake_list, "r") as f:
+        lines = f.readlines()
+
+    fault_names, r_counts = [], []
+    for ix, line in enumerate(lines):
+        line_l = line.split(" ")
+        fault_names.append(line_l[0].strip())
+
+        try:
+            r_counts.append(int(line_l[1].strip().rstrip("RrNnKk")))
+        except ValueError:
+            print(
+                "Failed to read line {} of cybershake list. Need to know"
+                "the number of realisations for each fault, quitting!".format(ix + 1)
+            )
+            # Raise the exception so execution stops
+            raise
+
+    return np.asarray(fault_names), np.asarray(r_counts)
+
+
+def get_new_progress_df(root_dir, runs_dir, faults, fault_names, r_counts):
+    """Gets a new progress dataframe, runs the full estimation + collects
+    all actual core hours.
+    """
+    # Run the estimation
+    est_args = argparse.Namespace(
+        vms_dir=sim_struct.get_VM_dir(root_dir),
+        sources_dir=sim_struct.get_sources_dir(root_dir),
+        runs_dir=runs_dir,
+        fault_selection=None,
+        root_yaml=None,
+        output=None,
+        verbose=False,
+        models_dir=None,
+    )
+    df = est_cybershake(est_args)
+    grouped_df = df.groupby("fault_name").sum()
+    grouped_df.sort_index(axis=0, level=0, inplace=True)
+
+    # Sanity check
+    if np.any(grouped_df.index.values != fault_names):
```

Comment:
Is this basically an assert? With assert, you can put a custom error on the same line rather than throwing an exception.